### PR TITLE
Liqun/ort module perf1

### PR DESCRIPTION
**Pull request by liqunfu**
**Description**: Log performance date.
Currently performance logging is done within onnxruntime/orttraining/orttraining/models/runner/training_runner.cc. To adapt to OrtModule, naturally the logging shall be done at the frontend. It allows us to log same performance data for training with both Pytorch and ORT so we can make direct apple-to-apple comparison. 
To log performance data at the frontend also eases code maintenance.
This is the phase 1 of performance logging work which simply make powerbi database connection and inserting performance data into the database. In the next phase we will finalize Bert and Huggingface training scripts and to put logging code into the training script.

**Motivation and Context**
- To replace manual perf run before devops pipeline

### CODE REVIEW

**Comment by SherlockNoMad** [comment]
Let's try to include label component:ortmodule in these PRs. Makes it easier to keep track of.

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    AvgTimePerBatch,\
+    Throughput,\
+    StabilizedThroughput,\
+    EndToEndThroughput,\
+    TotalTime,\
+    AvgCPU,\
+    Memory,\
+    RunConfig,\
+    Time)\
+VALUES\
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:75]: was the indentation intentionally done this way or was it unintentional?

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    completion_time = datetime.datetime.now()
+    perf_metrics_duration = completion_time - start_time
+
+    for attribute in required_attributes_for_perf_metrics:
+        if not hasattr(args, attribute):
+            raise ValueError('args does not contain all attributes needed to calculate perf metrics. \
+                Please prepare perf_metrics and call log_perf_metrics instead')
+
+    perf_metrics = {}    
+    perf_metrics['Model'] = args.model_name
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:152]: Can this lead to some kind of injection attacks?

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        TotalTime=perf_metrics['TotalTime'],
+        AvgCPU=perf_metrics['AvgCPU'],
+        Memory=perf_metrics['Memory'],
+        RunConfig=perf_metrics['RunConfig'],
+        Time=perf_metrics['Time'])
+    conn.cursor().execute(insert_table_script_values)
+    conn.commit()
+    conn.cursor().close()
+    conn.close()
+    print("Done.")
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:226]: This print statement can be a bit more descriptive?

File: tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-perf-test-ci-pipeline.yml [hunk 4]
```diff
@@ -22,17 +22,37 @@ jobs:
         --config RelWithDebInfo
         --enable_training
         --update --build
+        --build_wheel
         "
+        -u
```
Comment (baijumeswani) [inline tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-perf-test-ci-pipeline.yml:27]: ```-u``` is indicative of ```ORTModule``` build which installs the latest version of transformers library and (dependent on folder ```ortmodule``` being there).
I think master does not have all those changes, so this argument here is not necessary.

Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-perf-test-ci-pipeline.yml:27]: Nevermind, I was mistaken. Did not see the test being run was an ortmodule test.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_bert_classifier.py [hunk 5]
```diff
@@ -431,5 +441,43 @@ def main():
     print("  Accumulated training took:                {:.4f}s".format(total_training_time))
     print("  Accumulated validation took:              {:.4f}s".format(total_test_time))
 
+    if args.log_perf_metrics:
+        perf_metrics = {}
```
[the remark this answers is not in the document]
Comment (ytaous) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_bert_classifier.py:445]: +1

---
In reply to: [587694876](https://github.com/microsoft/onnxruntime/pull/6806#discussion_r587694876) [](ancestors = 587694876)

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_bert_classifier.py [hunk 6]
```diff
@@ -431,5 +441,43 @@ def main():
...
+        perf_metrics['Model'] = 'orttraining_test_ortmodule_bert_classifier'
+        perf_metrics['BatchId'] = 'na'
```
Comment (ytaous) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_bert_classifier.py:447]: assuming you'll use perf_metrics_start_time to publish this right?
this will be used for BI charts

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py (hunk 1 above)
Comment (ytaous) [reply orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:75]: perhaps we should indent the values below

---
In reply to: [587699060](https://github.com/microsoft/onnxruntime/pull/6806#discussion_r587699060) [](ancestors = 587699060)

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+if __name__ == '__main__':
+    args = parse_arguments()
+    conn = ConnectToPerfDashboardDb(args.mysql_server_name, args.power_bi_user_name, args.password, args.database)
+    conn.cursor().execute(create_table_script)
```
Comment (ytaous) [inline orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:231]: >create_table_script [](start = 26, length = 19)

would this failed if table already exists?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_bert_classifier.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (liqunfu) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_bert_classifier.py:445]: done

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py (hunk 2 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:152]: used parameter to guard against injection attacks. also made the code more precise

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py (hunk 3 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:226]: changed to: perf_metrics logged into power-bi database.

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py (hunk 7 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:231]: I put this code in so that we have the table definition in repo. otherwise if we lost the database, we lost it definition. You are correct that this is just a onetime thing.

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py (hunk 1 above)
Comment (liqunfu) [reply orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:75]: indent corrected

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_bert_classifier.py [hunk 8]
```diff
@@ -312,6 +312,49 @@ def format_time(elapsed):
...
+def bert_classifier_log_perf_metrics(args, train_dataloader, perf_metrics_start_time):
+    total_train_steps = 0
+    if args.train_steps > 0:
+        total_train_steps = args.train_steps * args.epochs
+    else:
+        total_train_steps = len(train_dataloader) * args.epochs
+
+    perf_metrics = {}
+    perf_metrics['Model'] = 'orttraining_test_ortmodule_bert_classifier'
+    perf_metrics['BatchId'] = 'na'
```
Comment (ytaous) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_bert_classifier.py:324]: >BatchId [](start = 18, length = 7)

just to clarify, the batchId is populated using datetime, I wonder how this field will be populated. Perhaps you can leverage datetime.datetime.now() again.

**Review by ytaous — APPROVED** [review]

File: orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+def parse_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--mysql_server_name', help='Perf dashboard mysql server name')
+    parser.add_argument('--power_bi_user_name', help='Power BI user name')
+    parser.add_argument('--password', help='password', default=None)
+    parser.add_argument('--database', help='The dashboard database')
+    return parser.parse_args()
+
+def ConnectAndInsertPerfMetrics(mysql_server_name, power_bi_user_name, password, database, perf_metrics):
+    conn = ConnectToPerfDashboardDb(mysql_server_name, power_bi_user_name, password, database)
```
Comment (kit1980) [inline orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:195]: Please follow PEP 8 style for function names: "Function names should be lowercase, with words separated by underscores as necessary to improve readability" (instead of `ConnectAndInsertPerfMetrics`
and `ConnectToPerfDashboardDb`).

Comment (liqunfu) [reply orttraining/orttraining/test/python/perf_log/ort_module_perf_test_tools.py:195]: name changed
